### PR TITLE
feat: improve deck creation/rename UX

### PR DESCRIPTION
**Pull request by david-allison**
## Purpose / Description

When a deck was created, there was little visual feedback on success/failure:

* When a deck was invalid, the dialog just closed
* When a deck was created/renamed, the user was not informed

## Fixes
* Fixes #10528
* Solves part of #11668: deck creation

## Approach
* toasts are used when a success occurs, 
* validation on the form is used if a deck couldn't be created
  * The OK button is greyed out 
  * A message is shown if the name is used

## How Has This Been Tested?

Screenshot 2022-08-23 at 23 15 19
Screenshot 2022-08-23 at 23 15 11
Screenshot 2022-08-23 at 23 14 58

## Checklist
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [x] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

**Comment by BrayanDSO** [comment]
I'd rather use Snackbars instead of toasts, as per Material design [recommendation](https://material.io/components/snackbars/android#using-snackbars)

> Android also provides a [Toast](https://developer.android.com/reference/android/widget/Toast.html) class with a similar API that can be used for displaying system-level notifications. Generally, snackbars are the preferred mechanism for displaying feedback messages to users, as they can be displayed in the context of the UI where the action occurred. Reserve Toast for cases where this cannot be done.

And on a future PR, or on this if you want to, an Undo button can be added

**Comment by david-allison** [comment]
I'd like a sample pattern of how the Dialog -> Snackbar relationship should be defined first.

Maybe via `Dialog.ownerActivity`?

**Comment by BrayanDSO** [comment]
> I'd like a sample pattern of how the Dialog -> Snackbar relationship should be defined first.

> Maybe via Dialog.ownerActivity, but this would cause code complexity if not handled carefully

Seems fine. Maybe `context` parameter could be replaced with activity, as activity serves as context too

**Comment by david-allison** [comment]
As defined currently, we have no `Activity`

**Comment by BrayanDSO** [comment]
WDYM? That AnkiDroid activities are all `AppCompatActivity`/`FragmentActivity` under the hood?

**Comment by david-allison** [comment]
We're in `CreateDeckDialog`, which has no specific reference to an `Activity`, only a `Context`

**Comment by BrayanDSO** [comment]
Oh, sorry, misunderstood that.  

My suggestion is to replace the `context: Context` parameter, with a `activity: Activity` parameter.

So we can use `activity` as context where `context` was used + use snackbars with `activity.showSnackbar()`

**Comment by david-allison** [comment]
I'll hold off on this and will add a TODO. This is because I'm unfamiliar with the standard Android pattern for handling this situation and architecturally I'd like to think about other options,

We may not want a hard dependency on `Activity` in the constructor

1. I'd like to move to a fragment-based app at some point in the future,
2. It holds a hard reference to an Activity context instead of allowing for an `Application` context
3. Requires a change from `Context` -> `Activity` in `DeckPickerFloatingActionMenu`, which ins't the end of the world

**Review by BrayanDSO — APPROVED** [review]

**Review by lukstbit — CHANGES_REQUESTED** [review]
The changes don't take in consideration creating sub decks with a name already in use. Scenario:

_I have decks A and B => create a sub deck in A named B => receive error the deck name is already used although it should permit the usage of the name_ 

Another scenario:

_I have deck A with sub deck A1 => attempt to create again a sub deck in A named A1 => no error visual indicator in the dialog that A1 already exists + also get toast with deck created_

**Comment by david-allison** [comment]
Good catch, will add the missing unit tests now a bug's been found.

File: AnkiDroid/src/test/java/com/ichi2/anki/dialogs/utils/MaterialDialogUtils.kt [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+package com.ichi2.anki.dialogs.utils
+
+import com.afollestad.materialdialogs.MaterialDialog
+import com.afollestad.materialdialogs.WhichButton
+import com.afollestad.materialdialogs.actions.getActionButton
+import com.afollestad.materialdialogs.input.getInputField
+
+val MaterialDialog.positiveButton get() = this.getActionButton(WhichButton.POSITIVE)
+
+var MaterialDialog.input
```
Comment (lukstbit) [inline AnkiDroid/src/test/java/com/ichi2/anki/dialogs/utils/MaterialDialogUtils.kt:26]: We already have a file containing extensions on MaterialDialog(although its name is not so great): https://github.com/ankidroid/Anki-Android/blob/main/AnkiDroid/src/main/java/com/ichi2/utils/MaterialBuilderUtil.kt

I think this extensions should go there.

**Review by lukstbit — CHANGES_REQUESTED** [review]
App behaves as expected, tests pass.
Just the small issue with the extensions placement and LGTM.

File: AnkiDroid/src/test/java/com/ichi2/anki/dialogs/utils/MaterialDialogUtils.kt (hunk 1 above)
Comment (david-allison) [reply AnkiDroid/src/test/java/com/ichi2/anki/dialogs/utils/MaterialDialogUtils.kt:26]: I would like these to be test-only functions, would you want the class renamed?

Comment (lukstbit) [reply AnkiDroid/src/test/java/com/ichi2/anki/dialogs/utils/MaterialDialogUtils.kt:26]: Sorry, I completely missed, when I reviewed, that this class is in the test source sets. So I'm happy as it is now, the name is perfectly fine.

**Review by lukstbit — APPROVED** [review]
Looks good!

**Review by BrayanDSO — APPROVED** [review]
LGTM. Pending solving conflicts

**Comment by BrayanDSO** [comment]
Good bot, I didn't pay attention to the label. Though it's under my impression that you only warned me after I merged it. I'll be more careful next time

**Comment by krmanik** [comment]
I think it is bug, there needs to do some check if PR merged or not. This needs to fix. I will update it.

**Pull request merged**